server/src/quotas.rs

```diff
@@ -0,0 +1,373 @@
+    organization_id: u64,
+) -> Result<Option<RetryAfter>, QuotasError> {
+    let timestamp = UnixTimestamp(
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .as_ref()
+            .map(Duration::as_secs)
+            .unwrap_or(0),
+    );
+
+    let mut invocation = IS_RATE_LIMITED.prepare_invoke();
+    let mut redis_quotas = vec![];
+
+    for quota in quotas {
+        match *quota {
+            Quota::RejectAll(RejectAllQuota { ref reason_code }) => {
+                // A zero-sized quota is the absolute worst-case. Do not call
+                // into Redis at all, and do not increment any keys, as one
+                // quota has reached capacity (this is how regular quotas behave
+                // as well).
+
+                return Ok(Some(RetryAfter {
+                    when: Instant::now() + Duration::from_secs(60),
+                    reason_code: reason_code.clone(),
+                }));
+            }
+            Quota::Redis(ref redis_quota) => {
+                let shift = organization_id % redis_quota.window;
+
+                let key = get_redis_key(redis_quota, timestamp, shift, organization_id);
+                let return_key = get_refunded_quota_key(&key);
+                invocation.key(key);
+                invocation.key(return_key);
+
+                let expiry = get_next_period_start(redis_quota, timestamp, shift);
+                let lua_quota = redis_quota.limit.map(|x| x as i64).unwrap_or(-1);
+                invocation.arg(lua_quota);
+                invocation.arg(expiry.0);
+                redis_quotas.push(redis_quota);
+            }
+        }
+    }
+
+    if redis_quotas.is_empty() {
+        return Ok(None);
+    }
+
+    let rejections: Vec<bool> = match redis_pool {
+        RedisPool::Cluster(ref pool) => {
+            let mut client = pool.get().map_err(QuotasError::RedisPool)?;
+            invocation
+                .invoke(&mut *client)
+                .map_err(QuotasError::Redis)?
+        }
+        RedisPool::Single(ref pool) => {
+            let mut client = pool.get().map_err(QuotasError::RedisPool)?;
+            invocation
+                .invoke(&mut *client)
+                .map_err(QuotasError::Redis)?
+        }
+    };
+
+    let mut worst_case = None;
+
+    for (quota, is_rejected) in redis_quotas.iter().zip(rejections) {
+        if !is_rejected {
+            continue;
+        }
+
+        let shift = organization_id % quota.window;
+        let delay = get_next_period_start(quota, timestamp, shift).0 - timestamp.0;
+
+        if worst_case
+            .as_ref()
+            .map(|(worst_delay, _)| delay > *worst_delay)
+            .unwrap_or(true)
+        {
+            worst_case = Some((delay, quota.reason_code.clone()));
+        }
+    }
+
+    Ok(worst_case.map(|(delay, reason_code)| RetryAfter {
+        when: Instant::now() + Duration::from_secs(delay),
+        reason_code,
+    }))
+}
+
+fn get_redis_key(
```

Comment:
Can we add a bunch of tests just to make sure that we don't regress with formatting keys?

Reply:
still need to do this

Reply:
Added a simple test, looks fine.